docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,279 @@
+	// the appropriate `rest.Storage` handler. For Create and Update, Validate is
+	// called after the handler's own `Validate` method is invoked. This gives the
+	// plugin the opportunity to validate the object after the handler has
+	// validated it, but before it is persisted. For Delete, Validate is called
+	// immediately before the handler's `Delete` method is invoked. If an error is
+	// returned, the request is marked as failed and no changes are persisted.
+	Validate(a Attributes) error
+
+  // Handles returns true if this admission controller can handle the given operation
+	// where operation can be one of CREATE, UPDATE, DELETE, or CONNECT
+	Handles(operation Operation) bool
+}
+```
+
+To be able to invoke admission plugin validation at the right time, we'll need
+to modify the various `rest.Storage` interfaces. For example,
+
+```go
+// Creater is an object that can create an instance of a RESTful object.
+type Creater interface {
+  // New returns an empty object that can be used with Create after request data has been put into it.
+  // This object must be a pointer type for use with Codec.DecodeInto([]byte, runtime.Object)
+  New() runtime.Object
+
+  // Create creates a new version of a resource.
+  Create(ctx api.Context, obj runtime.Object) (runtime.Object, error)
+}
+```
+
+would need to become
+
+```go
+// Creater is an object that can create an instance of a RESTful object.
+type Creater interface {
+  // New returns an empty object that can be used with Create after request data has been put into it.
+  // This object must be a pointer type for use with Codec.DecodeInto([]byte, runtime.Object)
+  New() runtime.Object
+
+  // BeforeCreate ensures that common operations for all resources are performed
+  // on creation. It is invoked after admission defaulting and authorization. It only returns
+  // errors that can be converted to api.Status. It returns nil if the object should be created.
+  BeforeCreate(ctx api.Context, obj runtime.Object) error
+
+  // Validate ensures that a resource is valid on creation. It is invoked after BeforeCreate.
+  Validate(ctx api.Context, obj runtime.Object) error
+
+  // Create creates a new version of a resource.
+  Create(ctx api.Context, obj runtime.Object) (runtime.Object, error)
+}
+```
+
+### Current
+
+[apiserver create flow](https://github.com/kubernetes/kubernetes/blob/ef0c9f0c5b8efbba948a0be2c98d9d2e32e0b68c/pkg/apiserver/resthandler.go#L333)
+1. decode incoming request
+1. admit.Admit
+1. r.Create (the items below are part of the generic registry store / strategy-based storage)
+  1. rest.BeforeCreate
+    1. strategy.PrepareForCreate
+    1. api.FillObjectMetaSystemFields
+    1. api.GenerateName
+    1. strategy.Validate
+    1. validation.ValidateObjectMeta
+    1. strategy.Canonicalize
+  1. e.Storage.Create
+  1. e.AfterCreate
+  1. e.Decorator
+
+### Proposed apiserver create flow
+
+1. decode incoming request
+1. admit.ApplyDefaults
+1. admit.Authorize
```

Comment:
I see no good reason why this cannot be done right before admit.Validate.

I would like users that send invalid resources to always get an validation error and not another admission error confusing the process.  So having validation always happen right after ApplyDefaults but before we Authorize/Validate/EnforceConstraints seems preferred.